optimization/docs/pgbart_improvements.md

```diff
@@ -35,77 +37,101 @@ As expected, the majority of the time is spent within loops of the `astep` metho
-
-If we re-write parts of PG-BART in Cython and or the entire codebase in Rust, we would get the benefits of AOT compilation. This would improve the speed of CPU bound loops and reduce memory usage.
+* JIT compiler translates bytecode into machine instructions at runtime, just before executing it.
+* Allows the JIT compiler to optimize the machine code using runtime information and only compile code paths that are actually executed.
 
+If we re-write parts of PG-BART in Cython and or the entire codebase in Rust, we would get the benefits of AOT compilation. This would improve the speed of loops and potentially reduce memory usage.
 
 ### Vectorization and parallelization
 
-Another area of improvement in PG-BART is vectorization and parallelization.
-
+Another area of improvement in PG-BART is vectorization and parallelization. It has been [shown](https://github.com/Gattocrucco/bartz) that BART can be vectorized resulting in a significant speedup. The linked repo above achieves vectorization (and parallelization) using JAX, and vectorizes the growing of trees. Furthermore, a JAX implementation can be ran on accelerators (e.g. GPUs/TPUs) as JAX uses an XLA compiler.
 
 ### Comparison
 
-| Framework | Rewrite? | Compilation | Impl. Difficulty | Vectorization | Parallelization | Accelerator Support |
-|-----------|----------|-------------|------------------|---------------|-----------------|---------------------|
-| JAX       | Yes      | JIT         | Medium           | `vmap`        | `pmap`          | Yes                 |
-| Cython    | Partial  | AOT         | Medium           | ?             | OpenMP          | ?                   |
-| Rust      | Yes      | AOT         | High             | ?             | Native?         | ?                   |            
+| Framework | Rewrite? | Impl. Difficulty | Compilation | Vectorization | Parallelization | Accelerator Support |
+|-----------|----------|------------------|-------------|---------------|-----------------|---------------------|
+| JAX       | Yes      | Medium           | JIT         | `vmap`        | `pmap`          | Yes                 |
+| Cython    | Partial  | Medium           | AOT         | ?             | OpenMP          | ?                   |
+| Rust      | Yes      | High             | AOT         | Native        | Native?         | ?                   |            
 
 #### JAX
 
 Rewriting the PG-BART codebase in JAX would allow:
-- JIT compilation
-- Automatic vectorization `vmap` and parallelization `pmap`
-- Ability to run BART on accelerators (e.g. GPU/TPUs) in addition to CPU as JAX uses the XLA compiler.
+* JIT compilation
+* Automatic vectorization `vmap` and parallelization `pmap`
+* Ability to run BART on accelerators (e.g. GPU/TPUs) in addition to CPU as JAX uses the XLA compiler.
 
 An implementation of BART in JAX already [exists](https://github.com/Gattocrucco/bartz).
 
 #### Cython
```

Comment:
Some time ago, I started the translate Trees to Cython Tress but run into many trouble with the pickle of the objects .

Could be interesting first trying to translate the method 'logp' instead.

Reply:
Ahh good to know. Out of curiosity, do you have this translation somewhere in a branch?

> Could be interesting first trying to translate the method 'logp' instead.

I agree. Although I need to first understand the intrinsics of the `pytensor_function`.

Reply:
I have in a local branch with a lot of .pyx files, testing different methods in numba, cython and python.. but I could tried to upload some of them in the weekend, if I upload right now probably you could understand much of that code and files.

Reply:
That would be greatly appreciated. Thank you!

Reply:
I uploaded some files of [prototypes](https://github.com/Grupo-de-modelado-probabilista/BART/tree/feature/optimization-cython/cython-start) with cython and [algorithms.pyx](https://github.com/francojloyola/pymc-bart/blob/feature/cython-start/pymc_bart/algorithms.pyx), in my case the thing that was the main problem with cython in tree.py was pickle of the classes within. As I recall, the issue was with NDArray and pickle. You should tried anyways. 

Probably if you want to optimize the memory use you should work on the Tree or ParticleTree

Reply:
Thanks a lot! Much appreciated! :)

Reply:
no problem, if you need something just let me know. Probably this month or the next one I will tried to improve the memory usage in the project, surely I will look towards the script you have done also

Reply:
@fjloyola great! I am currently undergoing a NumFOCUS project with Osvaldo to improve the speed and performance of PyMC-BART. In this project, I am Cythonizing the tree-structure and other time-consuming functions and or methods. 

So, if you also plan to do some work on this, just let me know so I can commit and document my work to avoid duplicate work 😄